proposals/0000-std-experimental.rst

```diff
@@ -0,0 +1,177 @@
+Add -experimental flag
+======================
+
+.. author:: Moritz Angermann, Julian Ospald
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/617>`_.
+.. sectnum::
+.. contents::
+
+Introduce a new compiler flag ``-experimental`` which guards experimental
+features. We distinguish between experimental and non-experimental features by
+their respective breaking change processes. Experimental features may introduce
+breaking changes without prior warning at any time. Non-experimental (stable)
+features on the other hand would need to follow a change evolution/deprecation
+process and could only under exceptional circumstances introduce breaking
+changes.
+
+Motivation
+----------
+As GHC continues to evolve, there's a need to introduce new features that may
+not be fully stabilized or might undergo significant changes in future releases.
+These experimental features can potentially introduce breaking changes without
+warning between different compiler versions. By introducing the
+``-experimental`` flag, developers can opt-in to use these features, fully
+aware of the potential risks. This ensures that features available without this
+flag are stable and will only have breaking changes with a proper deprecation
+cycle.
+
+What is _currently_ considered experimental can at best be found out by proxy
+via a `GHC User Guide's search for EXPERIMENTAL <https://downloads.haskell.org/ghc/latest/docs/users_guide/search.html?q=EXPERIMENTAL&check_keywords=yes&area=default>`_.
+
+
+Proposed Change Specification
+-----------------------------
+- Introduce a new compiler flag ``-experimental``.
+- Introduce a new compiler flag ``-no-experimental`` (default).
+- Features not behind the ``-experimental`` flag will have to adhere to a
+  deprecation cycle before introducing breaking changes.
+
+``-experimental`` is a superset, therefore ``-experimental`` could
+depend on code without ``-experimental``. The inverse does not hold.
+
+Semantically the last ``-[no-]experimental`` flag passed to GHC will win.
+
+Examples
+--------
+Consider a hypothetical experimental feature ``X``. Without the
+``-experimental`` flag, trying to use ``X`` would result in a compiler error.
+However, with the flag, the feature can be used, but with the understanding that
+it might change in future releases. ::
+
+  -- Without -experimental
+  ghc Main.hs  -- This will throw an error if Feature X is used
+
+  -- With -experimental
+  ghc -experimental Main.hs  -- This will compile successfully with Feature X
+
+  -- With -no-experimental
+  ghc -experimental -no-experimental Main.hs -- This will throw an error if Feature X is used
+
+  ghc -no-experimental Main.hs -package foo   -- This will throw an error if Feature X is used anywhere in Main.hs or the package foo.
```

Comment:
This implies a feature that's not listed in the change specification: that `-experimental` is transitive. That means in particular that we're going to have to record something about usage of `-experimental` in interface files, like Safe Haskell. Doable, but certainly worth calling out explicitly!

Reply:
@angerman Are you planning to address this?

The Proposed Change Specification is currently very brief. I think we need a lot more detail about the semantics of the flag: how does it relate to `-package` (and what is the impact of that design on Cabal packages)? Can it be overridden in an `OPTIONS_GHC` pragma?